### PR TITLE
 Fixed #32114 -- Subtest pickle issue with ParallelTestRunner

### PR DESCRIPTION
**Pull request by jonnyarnold**
[Ticket](https://code.djangoproject.com/ticket/32114)

When a subtest fails during a parallel test run, it gets pickled and sent back to the main thread to be added to the RemoteTestResult. For this to work, the TestCase needs to be pickleable. Often it is not, in which case the original error message is never displayed to the user. 

Instead of trying to pickle the entire TestCase, which can contain all sorts of weird and unexpected things, we can restrict the pickled fields to a subset of the original.

@JEphron submitted a pull request for this (#13599), but it has since become stale. I've incorporated the recommendation to use `__getstate__` over a separate object from the old pull request.

I briefly looked over the recommendation from [this comment](https://code.djangoproject.com/ticket/32114#comment:9) to render the exception to a string, but I think `unittest` expects a `_SubTest` object, rather than a serialized exception. This might not be true, so let me know if I've missed something here!

---

This is my first contirubtion to Django, so go easy on me 🙏 Thanks for all your comments in advance!

### CODE REVIEW

File: tests/test_runner/test_parallel.py [hunk 1]
```diff
@@ -50,6 +50,15 @@ def dummy_test(self):
             with self.subTest(index=i):
                 self.assertEqual(i, 1)
 
+    def pickle_error_test(self):
+        """A dummy test for testing TestCaseDTO."""
+        # ^ note: 3.6's implementation of TestCase.shortDescription()
+        #   doesn't strip whitespace before taking the first line of the
+        #   docstring, so the single-line docstring is part of this test
```
Comment (felixxm) [inline tests/test_runner/test_parallel.py:57]: The current development version of Django support Python 3.8+, so this note and workaround are unnecessary.

**Review by felixxm — COMMENTED** [review]
@jonnyarnold Thanks for this patch :+1:

File: django/test/testcases.py [hunk 2]
```diff
@@ -233,6 +233,17 @@ def __call__(self, result=None):
...
+    def __getstate__(self):
+        """Returns a pickleable copy of the test.
+
+        ParallelTestSuite requires that all TestCases are pickleable,
+        but subclasses may add unpickleable properties to objects.
+        """
+        return {
+            '_testMethodName': self._testMethodName,
+            '_testMethodDoc': self._testMethodDoc,
+        }
```
Comment (felixxm) [inline django/test/testcases.py:245]: I'm afraid that this approach is too restrictive. There are other required attributes e.g. `_cleanups`, `_type_equality_funcs`, or `_subtest`. You can try to apply the following diff:
```diff
diff --git a/tests/test_client_regress/tests.py b/tests/test_client_regress/tests.py
index 1f085a4a03..52c58ee584 100644
--- a/tests/test_client_regress/tests.py
+++ b/tests/test_client_regress/tests.py
@@ -1378,6 +1378,9 @@ class RequestFactoryStateTest(SimpleTestCase):
     def test_request_after_client_2(self):
         # This test is executed after the previous one
         self.common_test_that_should_always_pass()
+        with self.subTest('sth'):
+            self.x = memoryview(b'')
+            assert False
 
 
 @override_settings(ROOT_URLCONF='test_client_regress.urls')
```
and run:
```bash
./runtests.py test_client_regress
```

Comment (jonnyarnold) [reply django/test/testcases.py:245]: Thanks for the comment @felixxm 🙏

1. Could you explain where these required attributes are used? I can then build test cases to check they still work. Also, is this a complete list, or do you think there are others?
2. I'm not sure I understand the patch -- when I run it, I get the expected behaviour (it fails with an `AssertionError` with `assert False`, and passes with `assert True`). I must be missing something, sorry!

Comment (felixxm) [reply django/test/testcases.py:245]: > Thanks for the comment @felixxm pray
> 
> 1. Could you explain where these required attributes are used? I can then build test cases to check they still work. Also, is this a complete list, or do you think there are others?

These are attributes used by `unittest`'s [`TestCase`](https://github.com/python/cpython/blob/1d81fdc4c004511c25f74db0e04ddbbb8a04ce6d/Lib/unittest/case.py#L325-L404). I cannot guarantee that no other required attributes are dynamically added.

> 2. I'm not sure I understand the patch -- when I run it, I get behaviour I'd expect (it fails with an `AssertionError` with `assert False`, and passes with `assert True`). I must be missing something, sorry!

You need to run tests in parallel, e.g. `./runtests.py test_client_regress --parallel=2`

Comment (jonnyarnold) [reply django/test/testcases.py:245]: Thanks for the clarification @felixxm. I'm running into some problems with the `--parallel=2` flag, even on master:

```
(.env) tests ) ./runtests.py
Testing against Django installed in '/Users/jonny/work/django/django'
Creating test database for alias 'default'...
Creating test database for alias 'other'...
^CSystem check identified no issues (22 silenced).
.....................................................................................................


(.env) tests ) ./runtests.py --parallel=2
Testing against Django installed in '/Users/jonny/work/django/django' with up to 2 processes
Creating test database for alias 'default'...
Cloning test database for alias 'default'...
Cloning test database for alias 'default'...
Creating test database for alias 'other'...
Cloning test database for alias 'other'...
Cloning test database for alias 'other'...
System check identified no issues (22 silenced).
Process SpawnPoolWorker-2:
Process SpawnPoolWorker-1:
Traceback (most recent call last):
  File "/usr/local/Cellar/python@3.10/3.10.1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/local/Cellar/python@3.10/3.10.1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/local/Cellar/python@3.10/3.10.1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/pool.py", line 114, in worker
    task = get()
  File "/usr/local/Cellar/python@3.10/3.10.1/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/queues.py", line 368, in get
    return _ForkingPickler.loads(res)
  File "/Users/jonny/work/django/tests/generic_views/test_dates.py", line 9, in <module>
    from .models import Artist, Author, Book, BookSigning, Page
  File "/Users/jonny/work/django/tests/generic_views/models.py", line 7, in <module>
    class Artist(models.Model):
  File "/Users/jonny/work/django/django/db/models/base.py", line 108, in __new__
    app_config = apps.get_containing_app_config(module)
  File "/Users/jonny/work/django/django/apps/registry.py", line 253, in get_containing_app_config
    self.check_apps_ready()
  File "/Users/jonny/work/django/django/apps/registry.py", line 136, in check_apps_ready
    raise AppRegistryNotReady("Apps aren't loaded yet.")
django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.
```

This is on master. Is this the same error message that you're seeing? If not, do you know why I might be getting this error?

Comment (felixxm) [reply django/test/testcases.py:245]: Unfortunately, `--parallel` doesn't work on Windows. We're working on this, see #15421.

Comment (smithdc1) [reply django/test/testcases.py:245]: > Unfortunately, `--parallel` doesn't work on Windows. 

Or MacOs.

**Comment by felixxm** [comment]
@jonnyarnold Do you have time to keep working on this?

**Comment by felixxm** [comment]
Closing due to inactivity.

**Pull request closed without merging**